File: src/cli/output.ts
import { MenvError } from "../core/errors.ts";

export type OutputMode = "pretty" | "json";

export interface Io {
  stdout(text: string): void;
  stderr(text: string): void;
}

export const processIo: Io = {
  stdout: (s) => void process.stdout.write(s),
  stderr: (s) => void process.stderr.write(s),
};

// Test double: collects writes instead of printing.
export function memoryIo(): Io & { out: string[]; err: string[] } {
  const out: string[] = [];
  const err: string[] = [];
  return { out, err, stdout: (s) => void out.push(s), stderr: (s) => void err.push(s) };
}

export function resolveMode(flag: string | undefined, env: Record<string, string | undefined>): OutputMode {
  const m = flag ?? env.MENV_OUTPUT ?? "pretty";
  if (m !== "pretty" && m !== "json") {
    throw new MenvError("VALIDATION", `invalid output mode "${m}" (pretty | json)`);
  }
  return m;
}

// The uniform envelope (spec: output-modes contract). Success → stdout in both
// modes; errors → stdout as an envelope in json (machine-readable stream) but
// stderr in pretty (keeps $(menv get …) pipelines clean).
export function emitResult(io: Io, mode: OutputMode, result: unknown, pretty: string): void {
  if (mode === "json") {
    io.stdout(`${JSON.stringify({ ok: true, result })}\n`);
    return;
  }
  io.stdout(pretty.endsWith("\n") ? pretty : `${pretty}\n`);
}

export function emitError(io: Io, mode: OutputMode, e: MenvError): void {
  if (mode === "json") {
    io.stdout(`${JSON.stringify({ ok: false, error: { code: e.code, message: e.message, details: e.details ?? null } })}\n`);
    return;
  }
  io.stderr(`menv: ${e.message}\n`);
}

// For the top-level error handler: the parse may have failed, so sniff the raw
// argv instead of trusting commander. Unknown values fall back to pretty.
export function peekJsonMode(argv: string[], env: Record<string, string | undefined>): OutputMode {
  const i = argv.indexOf("--output");
  const fromArgv = i !== -1 ? argv[i + 1] : argv.find((a) => a.startsWith("--output="))?.slice("--output=".length);
  const m = fromArgv ?? env.MENV_OUTPUT;
  return m === "json" ? "json" : "pretty";
}

File: go/internal/cli/output.go
package cli

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/nikrabaev/menv/go/internal/core"
)

// OutputMode controls how results are emitted.
type OutputMode string

const (
	ModeJSON   OutputMode = "json"
	ModePretty OutputMode = "pretty"
)

// Io is the minimal write interface for CLI output.
type Io struct {
	Stdout func(text string)
	Stderr func(text string)
}

// ProcessIo returns an Io that writes to os.Stdout / os.Stderr.
func ProcessIo() Io {
	return Io{
		Stdout: func(s string) { fmt.Fprint(os.Stdout, s) },
		Stderr: func(s string) { fmt.Fprint(os.Stderr, s) },
	}
}

// ResolveMode converts the --output flag value to an OutputMode.
func ResolveMode(flag string) (OutputMode, error) {
	switch flag {
	case "", "pretty":
		return ModePretty, nil
	case "json":
		return ModeJSON, nil
	default:
		return "", &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("invalid output mode %q (pretty | json)", flag),
		}
	}
}

// EmitResult writes a result in the appropriate output mode.
func EmitResult(io Io, mode OutputMode, result any, pretty string) {
	if mode == ModeJSON {
		data, _ := json.Marshal(map[string]any{"ok": true, "result": result})
		io.Stdout(string(data) + "\n")
		return
	}
	if !strings.HasSuffix(pretty, "\n") {
		pretty += "\n"
	}
	io.Stdout(pretty)
}

// EmitError writes an error in the appropriate output mode.
func EmitError(io Io, mode OutputMode, e *core.MenvError) {
	if mode == ModeJSON {
		data, _ := json.Marshal(map[string]any{
			"ok": false,
			"error": map[string]any{
				"code":    string(e.Code),
				"message": e.Message,
				"details": e.Details,
			},
		})
		io.Stdout(string(data) + "\n")
		return
	}
	io.Stderr("menv: " + e.Message + "\n")
}

// PeekJSONMode inspects raw argv for --output=json before full parsing.
func PeekJSONMode(argv []string) OutputMode {
	for i, a := range argv {
		if a == "--output" && i+1 < len(argv) && argv[i+1] == "json" {
			return ModeJSON
		}
		if a == "--output=json" {
			return ModeJSON
		}
	}
	return ModePretty
}
